torch_xla/csrc/ops/index_op.cpp

```diff
@@ -0,0 +1,172 @@
+#include "torch_xla/csrc/ops/index_op.h"
+
+#include <ATen/ExpandUtils.h>
+#include <ATen/Functions.h>
+
+#include "tensorflow/compiler/xla/xla_client/debug_macros.h"
+#include "torch_xla/csrc/aten_xla_bridge.h"
+#include "torch_xla/csrc/lowering_context.h"
+#include "torch_xla/csrc/ops/arithmetic_ir_ops.h"
+#include "torch_xla/csrc/ops/infer_output_shape.h"
+#include "torch_xla/csrc/ops/ops.h"
+#include "torch_xla/csrc/tensor_util.h"
+#include "torch_xla/csrc/xla_lower_util.h"
+
+namespace torch_xla {
+namespace {
+
+void CheckIndexTensorTypes(at::TensorList indices, const Device* device) {
+  for (auto& tensor : indices) {
+    if (tensor.defined()) {
+      at::ScalarType scalar_type = tensor.scalar_type();
+      if (scalar_type != at::kLong &&
+          scalar_type != XlaTensorTypeMapping(at::kByte, device)) {
```

Comment:
@dlibenzi Because TPU is using `S32` to map `Byte`, we have to do this ugly thing. Let me know if you see a better way.

Reply:
Hold on one second ...

Reply:
How about this?

https://github.com/pytorch/xla/pull/356

Reply:
Yeah, probably a better option.

Reply:
Unfortunately it didn't help, still getting `kInt` here even with your patch applied. I'm looking into it more.

Reply:
Then it is worse, and I think I know what is going on.
When we manipulate tensors, we lose that info:

```
Byte a = 1;
Byte b = 2;
auto c = a + b;

type(c) == Int
```

I am not sure the correct solution is to propagate the logic type.
Even because that is a HUGE change.

Reply:
We need to preserve the logical type across calls into the bridge. Shouldn't `bridge::CreateXlaTensor` take the logical type?

Reply:
The XLATensor::Create(ir::Value, Device) should take c10::optional<at::ScalarType>.

Reply:
Let me think for a second before dropping int code ...

Reply:
Maybe for your particular example, for this indexing problem looks like `Create(const at::Tensor& tensor, const Device& device, bool requires_grad)` is the culprit.

Reply:
Essentially, for the majority of the cases besides "cast" APIs, we'd need something like:

static XLATensor CreateFrom(const XLATensor& input, ir::Value value);

Or the equivalent:

XLATensor CreateFrom(ir::Value value) const;

The API would then call the Create(ir::Value, Device, optional<ScalarType>);

Reply:
Let me sketch one in a diff, then you can do it.
Or I do it.
We could decouple this, if the change required by the index PR to deal with it, is not huge.

But if we have code that wants to infer types, like the index() one, we need to propagate them.
Hopefully this won't come back bite our arse later.

Reply:
To give the idea:

```
diff --git a/torch_xla/csrc/tensor.cpp b/torch_xla/csrc/tensor.cpp
index 6495d79..89fbb8a 100644
--- a/torch_xla/csrc/tensor.cpp
+++ b/torch_xla/csrc/tensor.cpp
@@ -538,8 +538,7 @@ XLATensor::ViewIrNode XLATensor::GetViewIrNode(View* view) {
 XLATensor XLATensor::add(const XLATensor& input, const XLATensor& other,
                          const at::Scalar& alpha) {
   ir::NodePtr constant = ir::ops::ScalarOp(alpha, other.shape());
-  return Create(input.GetIrValue() + other.GetIrValue() * constant,
-                input.GetDevice());
+  return input.CreateFrom(input.GetIrValue() + other.GetIrValue() * constant);
 }
 
 void XLATensor::add_(XLATensor& input, const XLATensor& other,
@@ -551,8 +550,7 @@ void XLATensor::add_(XLATensor& input, const XLATensor& other,
 XLATensor XLATensor::sub(const XLATensor& input, const XLATensor& other,
                          const at::Scalar& alpha) {
   ir::NodePtr constant = ir::ops::ScalarOp(alpha, other.shape());
-  return Create(input.GetIrValue() - other.GetIrValue() * constant,
-                input.GetDevice());
+  return input.CreateFrom(input.GetIrValue() - other.GetIrValue() * constant);
 }
 
 void XLATensor::sub_(XLATensor& input, const XLATensor& other,
@@ -1487,6 +1485,10 @@ XLATensor XLATensor::CreateView(
   return Create(std::make_shared<View>(std::move(shape), alias), GetDevice());
 }
 
+XLATensor XLATensor::CreateFrom(ir::Value ir_value) const {
+  return Create(std::move(ir_value), GetDevice(), dtype());
+}
+
 XLATensor XLATensor::log_softmax(const XLATensor& input, xla::int64 dim) {
   return Create(ir::MakeNode<ir::ops::LogSoftmax>(input.GetIrValue(), dim),
                 input.GetDevice());
diff --git a/torch_xla/csrc/tensor.h b/torch_xla/csrc/tensor.h
index 6a27f47..2891603 100644
--- a/torch_xla/csrc/tensor.h
+++ b/torch_xla/csrc/tensor.h
@@ -646,6 +646,10 @@ class XLATensor {
   XLATensor CreateView(
       tensorflow::gtl::ArraySlice<const xla::int64> output_size) const;
 
+  // Create a new XLA tensor with the same metadata of the input tensor, and the
+  // new IR value.
+  XLATensor CreateFrom(ir::Value ir_value) const;
+
   // Discards all the XLA and IR data, by making the ATEN tensor one the only
   // source for this XLA tensor. An error is generated if the XLA tensor does
   // not have ATEN tensors data.
```

Reply:
That way, if we ever have to change the metadata to be propagated, we have a single funnel point.

Reply:
To land this PR, it is sufficient to set the logical type for this constructor: https://github.com/pytorch/xla/blob/master/torch_xla/csrc/tensor.h#L600. Anything wrong about doing that? It looks like a bug fix to me.

Reply:
That makes sense indeed.
If that fixes your PR, good!
Let's go ahead with it.
I will take care of the other thing, because once we do operations of tensors, we lose the information.

Reply:
Ok, I'll do that and land this.

Regarding your sketch, it does make sense. As long as we have IR nodes at all sites which require doing this, it'll be fine. Basically, we need two things:

1. When ATen tensors enter our system, preserve their logical type. This is achieved by fixing https://github.com/pytorch/xla/blob/master/torch_xla/csrc/tensor.h#L600.
2. We have to preserve the types across operations, which is achieved through your proposal.

By virtue of structural induction, we should be fine in all cases.